Harden validations, report server version & deprecate TTL fields

nbougalis opened the pull request:
This PR contains three separate commits:

### Remove obsolete TTL fields
Proposals and validations include a TTL field. This was part of an earlier idea which never really materialized. The fields were never really used in practice (a TTL value is not set by default). This change simply makes this official: the fields are deprecated, should not be used and are not coming back.

Ripple engineers will be implementing the system outlined by @JoelKatz in this [post](https://www.xrpchat.com/topic/33075-suggestion-reduce-relaying/).

### Include version in validations

Currently, it is impossible to know what version of the software a given validator is running. This information is useful for server operators, for developing who are putting together dashboards, and the community at large.

This change adds a "packed" version in a validation message. The message will, by default, only be sent every 256 ledgers.

### Harden validations

Several changes are included in this PR.

1. Track how many validations a given server publishes for a given ledger. This should be precisely one. When such validations are detected, the server will now log a message to notify server operators of this issue.
2. The commit adds a 64-bit "cookie" in every validation which can be used to _attempt_ to decipher whether multiple validations are the result of accidental misconfiguration vs. outright byzantine behavior. It is important to note that a malicious validator operator can manipulate the cookie.
3. Adds the hash of the last ledger a validator considers to be fully validated in the validation message.

nbougalis commented:
(an unofficial lookover of 99f907b by @bachase would be awesome!)

MarkusTeufelberger commented:
> When such validations are detected, the server will now log a message to notify server operators of this issue.

Could we optionally also get a full log (or sqlite database, if you feel fancy) of all validation messages received? As far as I know, this data is currently only available on WebSocket streams.

nbougalis commented:
> > When such validations are detected, the server will now log a message to notify server operators of this issue.
> 
> Could we optionally also get a full log (or sqlite database, if you feel fancy) of all validation messages received? As far as I know, this data is currently only available on WebSocket streams.

We used to record validations in an SQL table, and we removed that functionality (see https://github.com/ripple/rippled/commit/c5a95f1eb5a5eedc7ece46ad26505209c1703486), because it was not really used by anyone/anything.

I think (and this is just my personal position) that the best option is to use the subscription and to index the received data separately. But would we, for example, report conflicting validations in the subscription stream?

Happy to discuss what to do here.

bachase commented on src/ripple/consensus/Validations.h:
```diff
@@ -167,7 +167,11 @@ enum class ValStatus {
     /// A validation violates the increasing seq requirement
-    badSeq
+    badSeq,
+    /// Multiple validations for the same sequence for same from multiple validators
```
Typo, extra `for same`?

bachase commented on src/ripple/consensus/Validations.h:
```diff
@@ -590,9 +606,45 @@ class Validations
+
+                // If the ledger hash matches, it must be a validation for
+                // the same ledger:
+                bool const same = (it->second.ledgerID() == val.ledgerID());
```
I'm unclear how these conditions wouldn't already be blocked by the `badSeq` one above. That is, the `enforcer` requires a strictly increasing sequence number seen by a validator. So I think it already filters these alternate outcomes.

bachase replied:
If you do want to check these, an option would be to move them above before `badSeq` returns.

carlhua replied:
Interesting comment here @bachase . I would argue that this is a valid check due to malicious/byzantine behavior. An attacker can increment the seq without changing the content. So this check can help us catch those byzantine behaviors is what I am thinking. I was thinking maybe the signature check would block the logic here but it seems to me that sfLedgerHash is just one field within the digest so a byzantine node can still sign the same validation with a different sequence and this logic here would catch it

bachase replied:
@carlhua but doesn't this check presume the two validations being compared have the same sequence number? They both come from the `bySequence_` map in [Line 614](https://github.com/ripple/rippled/pull/3291/files/31796963fa7ca6226751366b78d324775f0fe584..330f35b5f5712c758057fed0a962d5d831960ad8#diff-46f83c3cd15dd825b63068c0ab4dea00R616).

carlhua replied:
@bachase you are right!!  These checks would still guard against the scenarios depicted in the code right?
1. Same sequence, hashs match but cookies don't so we say its multiple
2. Same sequence, hashes don't match, we say its conflicting

nbougalis replied:
I sorted this out already. I this I misfolded some of my work before pushing. See updated commit.

bachase commented on src/ripple/consensus/Validations.h:
```diff
@@ -590,9 +606,45 @@ class Validations
+                        std::max(it->second.signTime(), val.signTime()) -
+                        std::min(it->second.signTime(), val.signTime());
+
+                    if (diff > std::chrono::minutes{10} &&
```
Consider using one of the other parameterized durations over this hardcoded 10 minutes.

bachase commented on src/ripple/app/consensus/RCLConsensus.cpp:
```diff
@@ -745,40 +752,60 @@ RCLConsensus::Adaptor::validate(RCLCxLedger const& ledger,
 
-    STValidation::FeeSettings fees;
-    std::vector<uint256> amendments;
+    auto v = std::make_shared<STValidation>(
```
:+1: Much cleaner

bachase commented on src/test/consensus/Validations_test.cpp:
```diff
@@ -1080,6 +1080,14 @@ class Validations_test : public beast::unit_test::suite
     }
 
+    void
+    testDifferentCookies()
```
I think testing this would be useful; I'm a little rusty on the test harness here but I think it should be feasible and can try to help.

nbougalis replied:
Appreciate the offer. I'll work with someone else to try and get these tests so that more people learn about consensus instead of having all the information locked up in the heads of three or four people.

seelabs replied:
I agree that this is important. I'm not sure it needs to be addressed before this is merged (that's up to Brad and Nik), but it would be good to assign this task to someone before this is merged, just so we don't lose track of the task.

pwang200 replied:
I will be happy to work on it, probably after the negative UNL network tests, if that is OK.

carlhua commented on src/ripple/app/consensus/RCLValidations.cpp:
```diff
@@ -191,9 +192,26 @@ handleNewValidation(Application& app,
         if(j.debug())
             dmp(j.debug(), to_string(outcome));
 
+        if (outcome == ValStatus::conflicting && j.fatal())
```
Couple thoughts here:
1. Consider this condition to be one of the criteria to be added to nUNL @pwang200  (future, but if we agree we should open an issue to track)
2. Consider this condition to be folded into peer selection/overlay enhancement - this is a byzantine behavior and we should consider disconnecting from this node or at least not connect to this node. Currently, this might not be a good idea but in the future when we are actively relaying validations from untrusted peers, we should consider keep records and potentially not connecting to these bad nodes. My argument here is that this allows us to eliminate bad nodes from the network.

carlhua commented on src/test/consensus/Validations_test.cpp:
```diff
@@ -1080,6 +1080,14 @@ class Validations_test : public beast::unit_test::suite
+    void
+    testDifferentCookies()
+    {
+        testcase("Multiple Validations");
```
I think it would be good to have more unit tests here. In addition, I would ask @manojsdoshi to consider to have automated tests for this PR to force the logic to execute on QE network. This should never happen naturally on the real network but it would be useful to see it in action in test.

mDuo13 commented:
I don't see any documentation impacts beyond the new amendment. Am I missing anything? Are there changes to the fields provided in any of the API subscription streams for example?

bachase commented on src/ripple/app/consensus/RCLValidations.cpp:
```diff
@@ -191,13 +193,30 @@ handleNewValidation(Application& app,
             dmp(j.debug(), to_string(outcome));
 
-        if (outcome == ValStatus::badSeq && j.warn())
+        if (outcome == ValStatus::conflicting && j.fatal())
```
Note that this checks `j.fatal()`, but logs below with `j.warn()`.  Same for line 208

bachase commented on src/ripple/consensus/Validations.h:
```diff
@@ -587,11 +603,46 @@ class Validations
+            {
+                // If the ledger hash matches, it must be a validation for
+                // the same ledger:
+                bool const same = (seqit->second.ledgerID() == val.ledgerID());
```
Consider renaming to `sameID` to make it clear what "same" thing you are referring to.

bachase commented on src/ripple/consensus/Validations.h:
```diff
@@ -587,11 +603,46 @@ class Validations
+                // Two validations for the same sequence but for different
+                // ledgers. This could be the result of misconfiguration
+                // but it can also mean a Byzantine validator.
+                if (!same)
```
If they have different ledger IDs, what guarantees they still have the same sequence number at this spot in the code? Do you also have to check `seqInserted`?

nbougalis replied:
I'll double-check this. Thanks.

gregtatcam commented on src/ripple/proto/ripple.proto:
```diff
@@ -210,9 +214,14 @@ message TMValidatorList
+    optional bool checkedSignature = 2 [deprecated = true];
+
+    // Number of hops traveled
+    optional uint32 hops = 3           [deprecated = true];
```
Could `hops` still be useful for the `reduce relaying` task? We can infer the preferred peer to receive TMValidation and TMPropose from as the ones with the smallest number of hops.

nbougalis commented:
I _think_ I've addressed all your concerns, except for improved unit test @bachase. I've avoided rebasing this to make reviewing easier; once you sign-off, I'll do it and force-push the branch.

In the meantime, I'm running a rebased version against the network.

nbougalis commented:
I'm going to rebase this against *b4* and squash.
